### PR TITLE
feat: wire Last.fm, Discogs, MusicBrainz for styles/moods with tag classification

### LINKED ISSUE #641: Wire Last.fm, Discogs, and MusicBrainz for styles/moods with tag classification
[mode: plan] 
[model: sonnet] 
[effort: medium] 

## Summary

Wire Last.fm, Discogs, and MusicBrainz as styles/moods providers with tag classification, and update the default provider priorities to reflect community-recognized source quality.

## Problem / Motivation

The styles and moods fields have limited provider coverage:
- **Styles**: only TheAudioDB and Discogs are in the priority list, but Discogs does not actually return styles from the artist endpoint (styles are on releases)
- **Moods**: only TheAudioDB is configured
- **Last.fm**: returns user-generated tags via `artist.getinfo`, but currently dumps ALL tags into the `Genres` field (line 222 of `lastfm.go`) with no classification into styles or moods

This means most artists have sparse or empty styles/moods data, even when multiple providers have relevant information.

### 1. Last.fm Tag Classification

Last.fm's `artist.getinfo` returns tags that are a mix of genres, styles, and moods. Instead of dumping all tags into `Genres`, classify them using a curated static map.

**Tag classifier** (new file `internal/provider/lastfm/classify.go`):
```go
// TagClass represents the classification of a Last.fm tag.
type TagClass int

const (
    TagClassGenre TagClass = iota
    TagClassStyle
    TagClassMood
    TagClassIgnore // descriptors like "seen live", "favorites"
)

// classificationMap maps normalized (lowercase) tag names to their class.
// Seeded from Last.fm's own tag taxonomy and cross-referenced with
// AllMusic, Discogs, and RateYourMusic style/mood taxonomies.
var classificationMap = map[string]TagClass{
    // Moods
    "melancholic":   TagClassMood,
    "melancholy":    TagClassMood,
    "chill":         TagClassMood,
    "energetic":     TagClassMood,
    "atmospheric":   TagClassMood,
    "dreamy":        TagClassMood,
    "aggressive":    TagClassMood,
    "uplifting":     TagClassMood,
    "dark":          TagClassMood,
    "happy":         TagClassMood,
    "sad":           TagClassMood,
    "intense":       TagClassMood,
    "mellow":        TagClassMood,
    "romantic":      TagClassMood,
    "haunting":      TagClassMood,
    "euphoric":      TagClassMood,
    "brooding":      TagClassMood,
    "nostalgic":     TagClassMood,
    "anxious":       TagClassMood,
    "peaceful":      TagClassMood,
    "angry":         TagClassMood,
    "bittersweet":   TagClassMood,
    "sombre":        TagClassMood,
    "somber":        TagClassMood,
    "eerie":         TagClassMood,

    // Styles (sub-genres / descriptors of approach)
    "progressive rock":     TagClassStyle,
    "shoegaze":             TagClassStyle,
    "trip-hop":             TagClassStyle,
    "lo-fi":                TagClassStyle,
    "lo fi":                TagClassStyle,
    "dream pop":            TagClassStyle,
    "post-punk":            TagClassStyle,
    "post punk":            TagClassStyle,
    "art rock":             TagClassStyle,
    "noise rock":           TagClassStyle,
    "math rock":            TagClassStyle,
    "krautrock":            TagClassStyle,
    "slowcore":             TagClassStyle,
    "sadcore":              TagClassStyle,
    "chamber pop":          TagClassStyle,
    "baroque pop":          TagClassStyle,
    "neo-psychedelia":      TagClassStyle,
    "space rock":           TagClassStyle,
    "stoner rock":          TagClassStyle,
    "doom metal":           TagClassStyle,
    "black metal":          TagClassStyle,
    "death metal":          TagClassStyle,
    "thrash metal":         TagClassStyle,
    "power metal":          TagClassStyle,
    "symphonic metal":      TagClassStyle,
    "nu metal":             TagClassStyle,
    "metalcore":            TagClassStyle,
    "post-metal":           TagClassStyle,
    "sludge metal":         TagClassStyle,
    "acid house":           TagClassStyle,
    "deep house":           TagClassStyle,
    "tech house":           TagClassStyle,
    "progressive house":    TagClassStyle,
    "minimal techno":       TagClassStyle,
    "acid techno":          TagClassStyle,
    "detroit techno":       TagClassStyle,
    "intelligent dance music": TagClassStyle,
    "idm":                  TagClassStyle,
    "glitch":               TagClassStyle,
    "vaporwave":            TagClassStyle,
    "synthwave":            TagClassStyle,
    "chillwave":            TagClassStyle,
    "witch house":          TagClassStyle,
    "dark ambient":         TagClassStyle,
    "drone":                TagClassStyle,
    "noise":                TagClassStyle,
    "power pop":            TagClassStyle,
    "jangle pop":           TagClassStyle,
    "twee pop":             TagClassStyle,
    "sunshine pop":         TagClassStyle,
    "bubblegum pop":        TagClassStyle,
    "electropop":           TagClassStyle,
    "synth-pop":            TagClassStyle,
    "synth pop":            TagClassStyle,
    "new wave":             TagClassStyle,
    "no wave":              TagClassStyle,
    "post-rock":            TagClassStyle,
    "post rock":            TagClassStyle,
    "emo":                  TagClassStyle,
    "screamo":              TagClassStyle,
    "hardcore":             TagClassStyle,
    "post-hardcore":        TagClassStyle,
    "midwest emo":          TagClassStyle,
    "skramz":               TagClassStyle,
    "britpop":              TagClassStyle,
    "madchester":           TagClassStyle,
    "acid jazz":            TagClassStyle,
    "smooth jazz":          TagClassStyle,
    "free jazz":            TagClassStyle,
    "bebop":                TagClassStyle,
    "hard bop":             TagClassStyle,
    "cool jazz":            TagClassStyle,
    "fusion":               TagClassStyle,
    "nu jazz":              TagClassStyle,
    "delta blues":          TagClassStyle,
    "chicago blues":        TagClassStyle,
    "electric blues":       TagClassStyle,
    "country blues":        TagClassStyle,
    "outlaw country":       TagClassStyle,
    "alt-country":          TagClassStyle,
    "americana":            TagClassStyle,
    "roots rock":           TagClassStyle,
    "garage rock":          TagClassStyle,
    "surf rock":            TagClassStyle,
    "psychedelic rock":     TagClassStyle,
    "grunge":               TagClassStyle,

    // Ignore (non-musical descriptors)
    "seen live":      TagClassIgnore,
    "favorites":      TagClassIgnore,
    "favourite":      TagClassIgnore,
    "love":           TagClassIgnore,
    "awesome":        TagClassIgnore,
    "favourite songs": TagClassIgnore,
    "beautiful":      TagClassIgnore,
    // ... expand as needed
}

// ClassifyTags splits Last.fm tags into genres, styles, and moods.
// Unclassified tags default to genres (safest bucket for unknown tags).
func ClassifyTags(tags []Tag) (genres, styles, moods []string) {
    for _, t := range tags {
        if t.Name == "" { continue }
        switch classificationMap[strings.ToLower(strings.TrimSpace(t.Name))] {
        case TagClassStyle:
            styles = append(styles, t.Name)
        case TagClassMood:
            moods = append(moods, t.Name)
        case TagClassIgnore:
            // skip
        default:
            genres = append(genres, t.Name)
        }
    }
    return
}
```

**Update `lastfm.go` GetArtist** to use classification:
```go
// Before (current):
for _, tag := range info.Tags.Tag {
    if tag.Name != "" {
        meta.Genres = append(meta.Genres, tag.Name)
    }
}

// After:
genres, styles, moods := ClassifyTags(info.Tags.Tag)
meta.Genres = genres
meta.Styles = styles
meta.Moods = moods
```

### 2. Discogs Release-Based Styles

Discogs stores genres and styles on releases, not on the artist endpoint. To get styles for an artist, fetch their releases and aggregate:

**New method in `internal/provider/discogs/discogs.go`:**
```go
// ArtistReleasesResponse wraps the paginated releases list.
type ArtistReleasesResponse struct {
    Releases []ArtistRelease `json:"releases"`
}

// ArtistRelease is a single release from the artist releases endpoint.
type ArtistRelease struct {
    ID    int    `json:"id"`
    Title string `json:"title"`
    Type  string `json:"type"`  // "master" or "release"
    Role  string `json:"role"`  // "Main", "Appearance", etc.
    Year  int    `json:"year"`
}

// ReleaseDetail contains genre/style info from a specific release.
type ReleaseDetail struct {
    Genres []string `json:"genres"`
    Styles []string `json:"styles"`
}

// GetArtistStyles fetches the artist's releases and aggregates styles.
// Only considers "Main" role releases to avoid noise from appearances.
// Returns the top N most common styles across the discography.
func (a *Adapter) GetArtistStyles(ctx context.Context, artistID string) ([]string, error) {
    path := fmt.Sprintf("/artists/%s/releases?sort=year&sort_order=desc&per_page=50", artistID)
    data, err := a.doGet(ctx, path)
    if err != nil { return nil, err }

    var resp ArtistReleasesResponse
    if err := json.Unmarshal(data, &resp); err != nil { return nil, err }

    // Count style occurrences across main releases
    styleCounts := make(map[string]int)
    for _, rel := range resp.Releases {
        if !strings.EqualFold(rel.Role, "Main") { continue }
        if rel.Type != "master" { continue } // prefer masters over individual pressings
        detail, err := a.getReleaseDetail(ctx, rel.ID)
        if err != nil { continue }
        for _, s := range detail.Styles {
            styleCounts[strings.TrimSpace(s)]++
        }
    }

    // Sort by frequency, return top styles
    return topN(styleCounts, 10), nil
}
```

**Integration into GetArtist:**
```go
func (a *Adapter) GetArtist(ctx context.Context, id string) (*provider.ArtistMetadata, error) {
    meta := mapArtist(detail)
    // Fetch styles from releases (secondary API call)
    styles, err := a.GetArtistStyles(ctx, id)
    if err != nil {
        a.logger.Warn("failed to fetch Discogs styles from releases", "error", err)
    } else {
        meta.Styles = styles
    }
    return meta, nil
}
```

**Rate limiting consideration:** This makes additional API calls per artist (1 for releases list + up to N for release details). The existing rate limiter (`provider.RateLimiterMap`) handles this, but the release detail calls should be capped (e.g., max 10 master releases sampled).

### 3. MusicBrainz Genre Tags as Styles (Lower Priority)

MusicBrainz returns genre tags that sometimes include style-level granularity (e.g., "shoegaze", "post-punk"). However, MusicBrainz genres are already the primary genre source. Adding it for styles would require similar classification logic to Last.fm.

**Recommendation:** Add MusicBrainz to the styles priority list at lowest priority. Map its genre tags through a classifier that identifies style-level tags:

```go
// In internal/provider/musicbrainz/musicbrainz.go
// After fetching genres, also populate styles with sub-genre tags
for _, tag := range artist.Tags {
    if isStyleTag(tag.Name) {
        meta.Styles = append(meta.Styles, tag.Name)
    }
}
```

The `isStyleTag()` function can reuse the same classification map from the Last.fm classifier (extract to a shared package `internal/provider/tagclass/`).

### 4. Updated Default Priorities

```go
func DefaultPriorities() []FieldPriority {
    return []FieldPriority{
        // ... existing fields unchanged ...
        {Field: "styles", Providers: []ProviderName{
            NameDiscogs,      // curated taxonomy, gold standard for style classification
            NameAudioDB,      // structured strStyle field, community-maintained
            NameLastFM,       // broad coverage via user tags, classified
            NameMusicBrainz,  // genre tags that include some style-level entries
        }},
        {Field: "moods", Providers: []ProviderName{
            NameAudioDB,      // structured strMood field, only provider with dedicated mood data
            NameLastFM,       // user tags classified as moods (broad but noisy)
        }},
    }
}
```

**Priority rationale:**
- **Discogs** (styles #1): The most respected, curated style taxonomy in the music community. Used as reference by record collectors, DJs, and catalogers worldwide. Distinction between genres and styles is a core Discogs feature.
- **TheAudioDB** (styles #2, moods #1): Structured `strStyle` and `strMood` fields. Community-maintained with editorial review. The only provider with a dedicated mood field.
- **Last.fm** (styles #3, moods #2): Broadest coverage via millions of user-generated tags. Noisy but extensive. Tag classification reduces noise. Good fallback when structured providers lack data.
- **MusicBrainz** (styles #4): Genre tags sometimes include style-level entries. Lower priority since its taxonomy is less structured for styles specifically.

### 5. Shared Tag Classification Package

Extract the classification map to a shared location so both Last.fm and MusicBrainz (and future providers) can reuse it:

```
internal/provider/tagclass/
    tagclass.go       // ClassifyTag(), classification map, TagClass type
    tagclass_test.go  // tests with known tags
```

### Testing

- **Last.fm classification tests**: verify known tags are correctly classified (table-driven)
- **Discogs release aggregation tests**: mock release list + details, verify style extraction
- **MusicBrainz style extraction tests**: verify genre-to-style classification
- **Priority integration tests**: verify orchestrator uses providers in correct order for styles/moods
- **Testdata fixtures**: add style/mood data to existing testdata JSON files (e.g., `artist_radiohead.json`)
- **Edge cases**: empty tags, unknown tags (default to genre), tags with whitespace, duplicate tags

## Acceptance Criteria

- [ ] Last.fm tags classified into genres, styles, and moods (not all dumped into genres)
- [ ] Shared tag classification package with curated map (~200+ entries)
- [ ] Discogs fetches styles from artist releases (aggregated from masters)
- [ ] Discogs rate limiting respected for additional release detail calls
- [ ] MusicBrainz genre tags filtered for style-level entries
- [ ] Default priorities updated: styles = Discogs > AudioDB > Last.fm > MusicBrainz
- [ ] Default priorities updated: moods = AudioDB > Last.fm
- [ ] Existing tests pass (no regression in genre fetching)
- [ ] New unit tests for tag classification with table-driven cases
- [ ] New integration tests for Discogs release-based style fetching
- [ ] Testdata fixtures updated with style/mood data

## Additional Context

- Last.fm provider: `internal/provider/lastfm/lastfm.go` (line 220-224: current tag handling)
- TheAudioDB provider: `internal/provider/audiodb/audiodb.go` (line 279-283: already wired)
- Discogs provider: `internal/provider/discogs/discogs.go` (line 276: mapArtist, no styles)
- Discogs types: `internal/provider/discogs/types.go` (no genre/style fields on ArtistDetail)
- Default priorities: `internal/provider/settings.go` (line 304-317)
- Orchestrator field routing: `internal/provider/orchestrator.go` (line 320-328: styles/moods cases)
- Quality validation: `internal/provider/quality.go` (line 59: styles/moods have no min length)

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary

- Add shared `tagclass` package (~200+ curated entries) to classify tags into genres, styles, moods, or ignore
- Last.fm: classify tags into genres/styles/moods instead of dumping all into genres
- Discogs: aggregate styles from artist master releases (capped at 10 masters for rate limiting)
- MusicBrainz: extract style-level entries from genre/tag lists (filtered for zero-vote noise)
- Updated default priorities: styles=Discogs>AudioDB>LastFM>MusicBrainz, moods=AudioDB>LastFM
- Updated scraper capabilities for Last.fm (styles+moods) and MusicBrainz (styles)

Closes #641

## Test plan

- [ ] `go test ./internal/provider/tagclass/...` -- table-driven classification tests
- [ ] `go test ./internal/provider/lastfm/...` -- tag classification in mapArtist (Bjork fixture)
- [ ] `go test ./internal/provider/discogs/...` -- release-based style aggregation (mock endpoints)
- [ ] `go test ./internal/provider/musicbrainz/...` -- style extraction (Portishead fixture)
- [ ] `go test ./... -count=1` -- no regressions (full suite passes)
- [ ] Manual: refresh an artist and verify styles/moods populated from multiple providers

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enhanced metadata enrichment: Styles are now extracted from Discogs releases, Last.fm, and MusicBrainz sources.
  * Moods metadata is now available from Last.fm tag classification.
  * Intelligent tag classification system automatically categorizes metadata into genres, styles, and moods.

* **Tests**
  * Added comprehensive test coverage for style extraction and tag classification across all providers.

### CODE REVIEW

File: internal/provider/discogs/discogs.go [hunk 1]
```diff
@@ -273,6 +285,119 @@ func (a *Adapter) doRequest(ctx context.Context, reqURL, token string) ([]byte,
...
+	const maxMasters = 10
+	var masterIDs []int
+	for _, rel := range releases {
+		if rel.Role == "Main" && rel.Type == "master" {
+			masterIDs = append(masterIDs, rel.ID)
+			if len(masterIDs) >= maxMasters {
+				break
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/provider/discogs/discogs.go]: Fixed in 44d7f68.

File: internal/provider/tagclass/tagclass_test.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		{"   ", TagClassIgnore},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.tag, func(t *testing.T) {
+			got := Classify(tt.tag)
+			if got != tt.want {
+				t.Errorf("Classify(%q) = %d, want %d", tt.tag, got, tt.want)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/provider/tagclass/tagclass_test.go]: Already addressed in 44d7f68.

File: internal/provider/discogs/discogs.go [hunk 3]
```diff
@@ -142,7 +143,18 @@ func (a *Adapter) getArtistByID(ctx context.Context, id string) (*provider.Artis
...
+	meta := mapArtist(&detail)
+
+	// Fetch styles from releases (secondary API calls).
+	styles, err := a.aggregateStyles(ctx, id, token)
+	if err != nil {
+		a.logger.Warn("failed to fetch Discogs styles from releases",
+			slog.String("artist_id", id), slog.String("error", err.Error()))
+	} else {
+		meta.Styles = styles
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/provider/discogs/discogs.go]: This is intentional best-effort behavior. Style aggregation is a secondary enrichment call -- if it fails for any reason (including context cancellation), the artist metadata is still returned with whatever data was already fetched from the primary endpoint. The warning log ensures visibility. This matches the pattern used elsewhere in the codebase for optional secondary fetches.
